libraries/AP_Scripting/examples/plane_ship_landing.lua

```diff
@@ -0,0 +1,427 @@
+
+--[[
+   parameter values which are auto-set on startup
+--]]
+local key_params = {
+   FOLL_ENABLE = 1,
+   FOLL_OFS_TYPE = 1,
+   FOLL_ALT_TYPE = 0,
+   FOLL_OFS_Z = 0
+}
+
+-- check key parameters
+function check_parameters()
+   for p, v in pairs(key_params) do
+      local current = param:get(p)
+      if not current then
+         gcs:send_text(0,string.format("Parameter %s not found", p))
+      end
+      if math.abs(v-current) > 0.001 then
+         param:set_and_save(p, v)
+         gcs:send_text(0,string.format("Parameter %s set to %.2f was %.2f", p, v, current))
+      end
+   end
+end
+
+-- update the pilots throttle position
+function update_throttle_pos()
+   local tpos
+   if not rc:has_valid_input() then
+      tpos = THROTTLE_LOW
+   else
+      local tchan = rc:get_channel(RCMAP_THROTTLE:get())
+      local tval = (tchan:norm_input_ignore_trim()+1.0)*0.5
+      if tval >= 0.45 then
+         tpos = THROTTLE_HIGH
+      elseif tval >= 0.2 then
+         tpos = THROTTLE_MID
+      else
+         tpos = THROTTLE_LOW
+      end
+   end
+   if tpos ~= throttle_pos then
+      reached_alt = false
+      if landing_stage == STAGE_HOLDOFF and tpos <= THROTTLE_MID then
+         gcs:send_text(0,"Descending for approach")
+         landing_stage = STAGE_DESCEND
+      end
+      if landing_stage == STAGE_DESCEND and tpos == THROTTLE_HIGH then
+         gcs:send_text(0,"Climbing for holdoff")
+         landing_stage = STAGE_HOLDOFF
+      end
+   end
+   throttle_pos = tpos
+end
+
+-- get landing airspeed
+function get_land_airspeed()
+   if TECS_LAND_ARSPD:get() < 0 then
+      return TRIM_ARSPD_CM:get() * 0.01
+   end
+   return TECS_LAND_ARSPD:get()
+end
+
+--[[
+  calculate stopping distance assuming we are flying at
+  TECS_LAND_ARSPD and are approaching the landing target from
+  behind. Take account of the wind estimate to get approach
+  groundspeed
+--]]
+function stopping_distance()
+   -- get the target true airspeed for approach
+   local tas = get_land_airspeed() * ahrs:get_EAS2TAS()
+
+   -- add in wind in direction of flight
+   local wind = ahrs:wind_estimate():xy()
+
+   -- rotate wind to be in approach frame
+   wind:rotate(-math.rad(target_heading + SHIP_LAND_ANGLE:get()))
+
+   -- ship velocity rotated to the approach frame
+   local ship2d = target_velocity:xy()
+   ship2d:rotate(-math.rad(target_heading + SHIP_LAND_ANGLE:get()))
+
+   -- calculate closing speed
+   -- use pythagoras theorem to solve for the wind triangle
+   local tas_sq = sq(tas)
+   local y_sq = sq(wind:y())
+   local closing_speed
+   if tas_sq >= y_sq then
+      closing_speed = math.sqrt(tas_sq - y_sq)
+   else
+      -- min 1 m/s
+      closing_speed = 1.0
+   end
+
+   -- include the wind in the direction of the ship
+   closing_speed = closing_speed + wind:x()
+
+   -- account for the ship velocity
+   closing_speed = closing_speed - ship2d:x()
+
+   -- calculate stopping distance
+   return sq(closing_speed) / (2.0 * Q_TRANS_DECEL:get())
```

Comment:
I would quite like to see a `quadplane:` method for this. I have been thinking about using a different calculation for tailsitters based on the pitch up rate. Tiltrotors also have a similar problem if they take a long time to tilt, that time can be longer than the time to stop. 

We could add that in the future of course.

Reply:
i think that can be added in future, and I agree a quadplane binding would be good